### DISCUSSION TITLE
Possibility to use an attribute as a variant?

### DISCUSSION THREAD

**Discussion by itzaks** [unocss/unocss#1589] [Q&A] [answered]
Would be really cool to be able to do something like this:

````pug
button(is-active="true" bg="red [is-active=true]:blue)
````
Or maybe

````pug
button(class="is-active" bg="red .is-active:blue)
````
Any way to achieve something like this? This way, the attribute could be flipped via vue/react and the style would reflect this state.

**Comment by itzaks** [comment]
Seems like [variantSelector](https://github.com/unocss/unocss/blob/b16db44fba15d5b684ef278948389b286cbb388f/packages/preset-mini/src/_variants/misc.ts#L4-L15) is kind of close to what I'm looking for, however I'm not really sure how to use it. Is it documented anywhere?

````ts
export const variantSelector: Variant = {
  name: 'selector',
  match(matcher) {
    const match = matcher.match(/^selector-\[(.+?)\][:-]/)
    if (match) {
      return {
        matcher: matcher.slice(match[0].length),
        selector: () => match[1],
      }
    }
  },
}
````

**Reply by zyyv** [reply]
You can check the test case.
https://github.com/unocss/unocss/blob/f375524d9bca3f2f8b445b322ec0fc3eb124ec3c/test/preset-uno.test.ts#L46-L52
https://github.com/unocss/unocss/blob/f375524d9bca3f2f8b445b322ec0fc3eb124ec3c/test/__snapshots__/preset-uno.test.ts.snap#L40-L50

**Comment by itzaks** [comment, marked as answer]
I fixed it by creating a custom variant like this:

````ts

const stateMatcher: Variant = {
  name: 'state',
  match(matcher) {
    const match = matcher.match(/^state-([_\d\w-]+)[:-]/)
    if (match) {
      return {
        matcher: matcher.slice(match[0].length),
        selector: s => `.${match[1]}${s}`,
      }
    }
  },
}
````

Opens the possibility to switch states easily in vue like this:

````pug
div(bg="red-400 state-is-active:green-400" :class="[isActive && 'is-active']")
````

**Reply by itzaks** [reply]
I also created something like this for switching on attributes:

````ts
const stateAttrMatcher: Variant = {
  name: 'state-attr',
  match(matcher) {
    const match = matcher.match(/^attr-([\d\w-]+)[:-]/)
    if (match) {
      return {
        matcher: matcher.slice(match[0].length),
        selector: s => `${s}[${match[1]}="true"]`,
      }
    }
  },
}
`````

Use like this
````pug
div(bg="red-400 attr-active:green-400" :active="isActiveState")
````
